From: Martin Schiller <ms@dev.tdt.de>
Date: Tue, 16 Dec 2025 12:33:48 +0100
Subject: [PATCH] mmc: sdhci-of-arasan: Add some quirks for Intel LGM eMMC

64-bit ADMA has some issue in CQE, so we disable it and use 32-bit
instead.

We also have to use capability register bit-63 for HS400 support
indication.

Signed-off-by: Martin Schiller <ms@dev.tdt.de>
---
 drivers/mmc/host/sdhci-of-arasan.c | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

--- a/drivers/mmc/host/sdhci-of-arasan.c
+++ b/drivers/mmc/host/sdhci-of-arasan.c
@@ -1540,9 +1540,18 @@ static const struct sdhci_arasan_clk_ops
 	.sampleclk_ops = &intel_lgm_sampleclk_ops,
 };
 
+static const struct sdhci_pltfm_data sdhci_lgm_emmc_pdata = {
+	.ops = &sdhci_arasan_cqe_ops,
+	.quirks = SDHCI_QUIRK_CAP_CLOCK_BASE_BROKEN,
+	.quirks2 = SDHCI_QUIRK2_PRESET_VALUE_BROKEN |
+		SDHCI_QUIRK2_CLOCK_DIV_ZERO_BROKEN |
+		SDHCI_QUIRK2_CAPS_BIT63_FOR_HS400 |
+		SDHCI_QUIRK2_BROKEN_64_BIT_DMA,
+};
+
 static struct sdhci_arasan_of_data intel_lgm_emmc_data = {
 	.soc_ctl_map = &intel_lgm_emmc_soc_ctl_map,
-	.pdata = &sdhci_arasan_cqe_pdata,
+	.pdata = &sdhci_lgm_emmc_pdata,
 	.clk_ops = &intel_lgm_clk_ops,
 };
 
